### PR TITLE
feat(smart-wallet): tryExitOffer also tries to reclaim

### LINKED ISSUE #9239: enhance smart wallet behavior when invitation creation fails
## What is the Problem Being Solved?

Currently smart-wallet lacks tests for failure to create invitation. This is a failure we want to handle to ensure we're handling invitations correctly in the smart wallet.

## Test Plan

Have test coverage for more edge cases.

## Upgrade Considerations

```[tasklist]
### Tasks
- [ ] https://github.com/Agoric/dapp-offer-up/pull/76
- [ ] https://github.com/Agoric/agoric-sdk/pull/9250
- [ ] https://github.com/Agoric/agoric-sdk/pull/9254
- [ ] https://github.com/Agoric/agoric-sdk/pull/9204
- [ ] https://github.com/Agoric/agoric-sdk/pull/9240
- [ ] https://github.com/Agoric/agoric-sdk/pull/9236
- [ ] https://github.com/Agoric/agoric-sdk/pull/9238
```

### PR DESCRIPTION
**Pull request by dckc**
closes: #9239
~~stacked on #9236~~

## Description

 - [x] feat(smart-wallet): tryExitOffer reclaims withdrawn payments
   - [x] logs reclaimed balances
 - [x] upgrade core-eval, app.go
 - [x] feat: tryExitOffer CLI tool
 - [x] a3p-integration test: bad offer in prepare; test exitOffer tool in test phase

### Security / Scaling Considerations

Using `tryExitOffer` to trigger reclaim results in an irrelevant diagnostic that's possibly misleading:

```
2024-04-18T21:32:20.809Z SwingSet: ls: v43: Error#1: key bad-invitation-15 not found in collection live offer seats
```

### Documentation Considerations

the `exitOffer.js` tool is in sort of a strange place

### Testing Considerations

e2e test coverage seems good.

The unit test from #9236 doesn't cover this feature well because it doesn't run in the before-upgrade world

### Upgrade Considerations

needs cherry-picking

### CODE REVIEW

File: packages/builders/scripts/smart-wallet/build-wallet-factory3-upgrade.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  harden({
+    sourceSpec:
+      '@agoric/smart-wallet/src/proposals/upgrade-wallet-factory3-proposal.js',
+    getManifestCall: [
+      getManifestForUpgradeWallet.name,
+      {
+        walletRef: publishRef(
+          install('@agoric/smart-wallet/src/walletFactory.js'),
+        ),
+        zcfRef: publishRef(install('@agoric/zoe/src/contractFacet/vatRoot.js')),
```
Comment (Chris-Hibbert) [inline packages/builders/scripts/smart-wallet/build-wallet-factory3-upgrade.js]: No longer used?

File: packages/smart-wallet/src/proposals/upgrade-wallet-factory3-proposal.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
+// @ts-check
+import { E } from '@endo/far';
+
+const { keys } = Object;
+
+/**
+ * @param {BootstrapPowers & NonNullChainStorage} powers
+ * @typedef {PromiseSpaceOf<{ chainStorage: StorageNode }>} NonNullChainStorage
+ * @param {object} options
+ * @param {{ walletRef: VatSourceRef, zcfRef: VatSourceRef }} options.options
```
Comment (Chris-Hibbert) [inline packages/smart-wallet/src/proposals/upgrade-wallet-factory3-proposal.js]: don't need zcfRef anymore?

File: a3p-integration/proposals/a:upgrade-next/exit-reclaim.test.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
+import test from 'ava';
+import { execFile } from 'node:child_process';
+import { promisify } from 'node:util';
+
+const offerId = 'bad-invitation-15'; // cf. prepare.sh
+const from = 'gov1';
+
+test.before(async t => {
+  const $ = async (file, args) => {
```
Comment (Chris-Hibbert) [inline a3p-integration/proposals/a:upgrade-next/exit-reclaim.test.js]: Why does this have a name which is a symbol rather than a descriptive string?

**Review by Chris-Hibbert — CHANGES_REQUESTED** [review]

File: packages/builders/scripts/smart-wallet/build-wallet-factory3-upgrade.js (hunk 1 above)
Comment (dckc) [reply packages/builders/scripts/smart-wallet/build-wallet-factory3-upgrade.js]: shoot! that's right.

File: a3p-integration/proposals/a:upgrade-next/exit-reclaim.test.js (hunk 3 above)
Comment (dckc) [reply a3p-integration/proposals/a:upgrade-next/exit-reclaim.test.js]: it's patterned after zx and execa. maybe not a good enough reason

Comment (dckc) [reply a3p-integration/proposals/a:upgrade-next/exit-reclaim.test.js]: ok if I leave it?

Comment (turadg) [reply a3p-integration/proposals/a:upgrade-next/exit-reclaim.test.js]: not a blocker but I'd prefer we import `execa` instead of approximating it, so that it works consistently with all the other uses of it.

I also don't think we need to confine this helper to `context`. it's a test

File: a3p-integration/proposals/a:upgrade-next/exit-reclaim.test.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    const { stdout } = await promisify(execFile)(file, args);
+    return stdout.replace(/\n$/, '');
+  };
+  Object.assign(t.context, { $ });
+});
+
+test('exitOffer tool reclaims stuck payment', async t => {
+  const { $ } = t.context;
+
+  const addr = await $('agd', [
```
Comment (turadg) [inline a3p-integration/proposals/a:upgrade-next/exit-reclaim.test.js]: a3p lib has an agd helper, for the future. it even eschews ambient authority
https://github.com/Agoric/agoric-3-proposals/blob/3faecbf11b95da6044ebacd067c5b72f9a45f88f/packages/synthetic-chain/src/lib/core-eval.ts#L57-L60

File: a3p-integration/proposals/a:upgrade-next/exit-reclaim.test.js (hunk 3 above)
Comment (Chris-Hibbert) [reply a3p-integration/proposals/a:upgrade-next/exit-reclaim.test.js]: I would appreciate a comment, but I won't insist  on it. It didn't take me long to track the reference back to here.

File: a3p-integration/proposals/a:upgrade-next/prepare.sh [hunk 5]
```diff
@@ -1,9 +1,29 @@
 #!/bin/bash
 
 # Exit when any command fails
-set -e
+set -uxeo pipefail
```
Comment (turadg) [inline a3p-integration/proposals/a:upgrade-next/prepare.sh:4]: is this what we should use for all a3p scripts?

**Review by Chris-Hibbert — APPROVED** [review]

File: packages/smart-wallet/src/smartWallet.js [hunk 6]
```diff
@@ -1052,6 +1057,19 @@ export const prepareSmartWallet = (baggage, shared) => {
...
+            .tryReclaimingWithdrawnPayments(offerId)
+            .catch(e => {
+              facets.helper.logWalletError(
+                'recovery failed reclaiming payments',
+                e,
+              );
+              return [];
+            });
+          if (amts.length > 0) {
+            facets.helper.logWalletError('reclaimed', amts, 'from', offerId);
```
Comment (turadg) [inline packages/smart-wallet/src/smartWallet.js]: not a blocker but this is success, not an error.
```suggestion
            facets.helper.logWalletInfo('reclaimed', amts, 'from', offerId);
```

**Review by turadg — APPROVED** [review]

File: packages/smart-wallet/src/smartWallet.js (hunk 6 above)
Comment (Chris-Hibbert) [reply packages/smart-wallet/src/smartWallet.js]: Thank you, @turadg. I wondered why it was an error, but didn't get as far as identifying that it didn't need to be.

File: packages/smart-wallet/src/proposals/upgrade-wallet-factory3-proposal.js [hunk 7]
```diff
@@ -0,0 +1,70 @@
+// @ts-check
+import { E } from '@endo/far';
+
+const { keys } = Object;
```
Comment (mhofman) [inline packages/smart-wallet/src/proposals/upgrade-wallet-factory3-proposal.js]: As the bots say, unused

File: packages/smart-wallet/src/proposals/upgrade-wallet-factory3-proposal.js
Comment (mhofman) [inline packages/smart-wallet/src/proposals/upgrade-wallet-factory3-proposal.js:1]: Conceptually, is there a reason to use a new proposal file instead of tweaking the `upgrade-wallet-factory2-proposal.js` one which does functionally the same thing?

File: packages/smart-wallet/src/smartWallet.js [hunk 8]
```diff
@@ -928,10 +929,14 @@ export const prepareSmartWallet = (baggage, shared) => {
                 const purseP = facets.helper.purseForBrand(b);
 
                 // Now send it back to the purse.
-                return E(purseP).deposit(p);
+                return E(purseP)
+                  .deposit(p)
+                  .then(amt => out.push(amt));
```
Comment (mhofman) [inline packages/smart-wallet/src/smartWallet.js]: Nit, we don't need the return value
```suggestion
                  .then(amt => { out.push(amt); });
```

File: packages/smart-wallet/src/smartWallet.js [hunk 9]
```diff
@@ -1052,6 +1057,19 @@ export const prepareSmartWallet = (baggage, shared) => {
          * @throws if the seat can't be found or E(seatRef).tryExit() fails.
          */
         async tryExitOffer(offerId) {
+          const { facets } = this;
+          const amts = await facets.payments
+            .tryReclaimingWithdrawnPayments(offerId)
+            .catch(e => {
+              facets.helper.logWalletError(
+                'recovery failed reclaiming payments',
```
Comment (mhofman) [inline packages/smart-wallet/src/smartWallet.js:1067]: recovery ? This method says "Exit offer"

File: packages/smart-wallet/src/smartWallet.js (hunk 8 above)
Comment (mhofman) [reply packages/smart-wallet/src/smartWallet.js]: But as discussed below

```suggestion
                  .then(amt => {
                    out.push(amt);
                    brandPaymentRecord.delete(b);
                  });
```

File: packages/smart-wallet/src/smartWallet.js [hunk 10]
```diff
@@ -928,10 +929,14 @@ export const prepareSmartWallet = (baggage, shared) => {
                 const purseP = facets.helper.purseForBrand(b);
 
                 // Now send it back to the purse.
-                return E(purseP).deposit(p);
+                return E(purseP)
+                  .deposit(p)
+                  .then(amt => out.push(amt));
               }),
             );
+            return harden(out);
```
Comment (mhofman) [inline packages/smart-wallet/src/smartWallet.js:939]: Should `brandPaymentRecord` stick around after this? All payments have been deposited, no?

Which makes me wonder what happens if some payments fail to deposit but not other.

I would think the following would be more robust:
- For each payment successfully deposited, remove from `brandPaymentRecord`
- If all payments have been successfully deposited, remove `brandPaymentRecord` from `liveOfferPayments`

```suggestion
            liveOfferPayments.delete(offerId);
            return harden(out);
```

This is of course assuming no other code path will mutate the `brandPaymentRecord` while depositing the payments, which I believe is the case as only `withdrawGive` seem to do that.

If handling partial successful deposits, maybe this method should return a `{amounts: Amount[], error?: Error | null}`, and never reject?

**Review by mhofman — COMMENTED** [review]
Mostly a question about the existing behavior of `tryReclaimingWithdrawnPayments`

File: a3p-integration/proposals/a:upgrade-next/exitOffer.js
Comment (gibson042) [inline a3p-integration/proposals/a:upgrade-next/exitOffer.js:1]: This _is_ in a strange place. I think it would make more sense in [bin](https://github.com/Agoric/agoric-sdk/tree/master/bin), but I'm also open to alternatives.

File: a3p-integration/proposals/a:upgrade-next/exitOffer.js [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+  env.PATH = `${bin}:${env.PATH}`;
+  const action = { method: 'tryExitOffer', offerId };
+  const out = await doAction(action, from);
+  console.log(out);
+};
+
+main(process.argv, process.env).catch(e => {
+  if (e.name === 'USAGE' || e.code === 'ERR_PARSE_ARGS_UNKNOWN_OPTION') {
+    console.error(e.message);
+    process.exit(1);
```
Comment (gibson042) [inline a3p-integration/proposals/a:upgrade-next/exitOffer.js]: Redundant exit.
```suggestion
```

File: a3p-integration/proposals/a:upgrade-next/exitOffer.js [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+  );
+  await writeFile(tmpOffer, q(toCapData(action)));
+
+  const out = await showAndRun('agoric', [
+    'wallet',
+    ...flags({ 'keyring-backend': 'test' }),
+    'send',
+    ...flags({ offer: tmpOffer, from }),
+  ]);
+  return out.stdout;
```
Comment (gibson042) [inline a3p-integration/proposals/a:upgrade-next/exitOffer.js]: What is "tempfile"? If we're shelling out for this, wouldn't `mktemp` have the broadest support? But better still would be to use [built-in Node.js functionality](https://nodejs.org/docs/latest-v18.x/api/fs.html#fspromisesmkdtempprefix-options)—and to clean up after ourselves when successful.
```suggestion
  const tmpDir = await mkdtemp();
  const tmpOffer = path.join(tmpDir, 'offer.json');
  await writeFile(tmpOffer, q(toCapData(action)));

  const out = await showAndRun('agoric', [
    'wallet',
    ...flags({ 'keyring-backend': 'test' }),
    'send',
    ...flags({ offer: tmpOffer, from }),
  ]);
  await rm(tmpDir, { recursive: true, force: true }).catch(err => console.error(err));
  return out.stdout;
```

**Review by gibson042 — COMMENTED** [review]
Some comments, unfortunately superficial for lack of sufficiently deep smart wallet knowledge.

File: a3p-integration/proposals/a:upgrade-next/exit-reclaim.test.js (hunk 3 above)
Comment (dckc) [reply a3p-integration/proposals/a:upgrade-next/exit-reclaim.test.js]: > prefer we import `execa`

ok.

I had just finished the `exitOffer.js` CLI tool where a single file that uses only node's standard library would simplify installation / deployment and my mind was still in that mode.

File: a3p-integration/proposals/a:upgrade-next/prepare.sh (hunk 5 above)
Comment (dckc) [reply a3p-integration/proposals/a:upgrade-next/prepare.sh:4]: I got it from [The JISH Stack](https://kriskowal.com/jish/):

> This preamble will save you a lot of trouble.
>
> ```
> #!/bin/bash
> set -ueo pipefail
> IFS=$'\t\n'
> ```
>
> Use [shellcheck](https://www.shellcheck.net/).

I skipped IFS because I'm not doing any funky quoting. I did use `shellcheck` to be sure.

File: packages/smart-wallet/src/proposals/upgrade-wallet-factory3-proposal.js
Comment (dckc) [reply packages/smart-wallet/src/proposals/upgrade-wallet-factory3-proposal.js:1]: no. good catch.

The 2 wrinkles that motivated it (zcf, auto exit) are gone now.

File: a3p-integration/proposals/a:upgrade-next/exit-reclaim.test.js (hunk 4 above)
Comment (dckc) [reply a3p-integration/proposals/a:upgrade-next/exit-reclaim.test.js]: weird! importing `@agoric/synthetic-chain` hangs!

I put a `console.log()` at the top and bottom. I can see the top one but not the bottom one.

btw... the generated JS in synthetic-chain's dist is a _long_ way from just erasing the types. `const` is turned to `var`, and the modules are all concatenated without any namespace isolation, and there are `import` declarations in the middle of the statements.

File: a3p-integration/proposals/a:upgrade-next/exitOffer.js
Comment (turadg) [reply a3p-integration/proposals/a:upgrade-next/exitOffer.js:1]: I figured it was only meant for testing this proposal. If we want to publish the functionality, it belongs in `agoric-cli`. e.g. `agoric wallet  propose-exit`. But that's out of scope so I've filed https://github.com/Agoric/agoric-sdk/issues/9259

thanks for raising because we already have https://github.com/Agoric/agoric-sdk/blob/e1e0b2f06df1ebee2b23c1be23503f8d76e7a223/packages/agoric-cli/src/commands/inter.js#L469-L480

I'll make a suggestion the call site for this to use that instead.

File: a3p-integration/proposals/a:upgrade-next/exit-reclaim.test.js [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+    while (block < start + n) {
+      console.log('...block', block, '<', start + n);
+      await new Promise(r => setTimeout(r, 1000));
+      block = await getBlock();
+    }
+  };
+
+  const before = await getBalance();
+  t.log(from, { before });
+  await $('node', ['./exitOffer.js', '--id', offerId, '--from', from]);
```
Comment (turadg) [inline a3p-integration/proposals/a:upgrade-next/exit-reclaim.test.js]:
```suggestion
  // UNTIL https://github.com/Agoric/agoric-sdk/issues/9259
  await $('agops', ['inter', 'bid', 'cancel', offerId, '--from', from]);
```

File: packages/smart-wallet/src/smartWallet.js (hunk 9 above)
Comment (dckc) [reply packages/smart-wallet/src/smartWallet.js:1067]: In Apr 15 discussion with @turadg and @Chris-Hibbert, we agreed to expand `tryExitOffer` to include reclaiming rather than adding a new `tryReclaimOffer` method.

File: packages/smart-wallet/src/smartWallet.js (hunk 10 above)
Comment (dckc) [reply packages/smart-wallet/src/smartWallet.js:939]: re-visiting that logic is beyond the scope of this PR

IIUC, the `brandPaymentRecord` is deleted in `updateStatus`.

File: a3p-integration/proposals/a:upgrade-next/exitOffer.js
Comment (dckc) [reply a3p-integration/proposals/a:upgrade-next/exitOffer.js:1]: I'm not sure that bin directory is visible from the container where this test runs.

File: a3p-integration/proposals/a:upgrade-next/exitOffer.js (hunk 12 above)
Comment (dckc) [reply a3p-integration/proposals/a:upgrade-next/exitOffer.js]: Thanks.
TIL: node has built-in temp file support.

In agoric-sdk, we depend on various packages for this, so I inferred that node doesn't include it.

File: a3p-integration/proposals/a:upgrade-next/exit-reclaim.test.js (hunk 13 above)
Comment (dckc) [reply a3p-integration/proposals/a:upgrade-next/exit-reclaim.test.js]: Do we know that `agops` is in `$PATH`? from the shell, it seems to be an alias. trying it out... nope

```
    message: `Command failed with ENOENT: agops inter bid cancel bad-invitation-15 --from gov1 --keyring-backend=test␊
```

Comment (dckc) [reply a3p-integration/proposals/a:upgrade-next/exit-reclaim.test.js]: no joy; that command checks that the offer is live:

```
bad-invitation-15 not in live offer ids: bid-1713384417025
```

Comment (turadg) [reply a3p-integration/proposals/a:upgrade-next/exit-reclaim.test.js]: Ok. We could change that throw to a warn but given that you already have `exitOffer.js` working I agree it's fine to proceed with it.

Comment (turadg) [reply a3p-integration/proposals/a:upgrade-next/exit-reclaim.test.js]: Ok. We could change that throw to a warn but given that you already have `exitOffer.js` working I agree it's fine to proceed with it.

File: a3p-integration/proposals/a:upgrade-next/exit-reclaim.test.js (hunk 4 above)
Comment (Chris-Hibbert) [reply a3p-integration/proposals/a:upgrade-next/exit-reclaim.test.js]: > weird! importing @agoric/synthetic-chain hangs!

This was blocking me for most of the week. Even though I had done it myself more than once, after @turadg suggested running `yarn install` inside the container, the problem cleared up.

File: packages/smart-wallet/src/smartWallet.js (hunk 10 above)
Comment (mhofman) [reply packages/smart-wallet/src/smartWallet.js:939]: I don't think updateStatus triggers in the path where reclaim payment does. But I may not fully understand the plumbing.

Comment (dckc) [reply packages/smart-wallet/src/smartWallet.js:939]: right; updateStatus doesn't trigger until later. but it seems soon enough

Comment (mhofman) [reply packages/smart-wallet/src/smartWallet.js:939]: My claim is that it never triggers since there is no seat or offer to exit in this case.
That means we'll just be leaving junk around in state

Comment (dckc) [reply packages/smart-wallet/src/smartWallet.js:939]: OK, I see now.

IOU an issue.
